job-server/src/spark.jobserver/JobCache.scala

```diff
@@ -22,28 +26,51 @@ class JobCacheImpl(maxEntries: Int,
                    loader: ContextURLClassLoader) extends JobCache {
   import scala.concurrent.duration._
 
-  private val cache = new LRUCache[(String, DateTime, String), JobJarInfo](maxEntries)
+  private val cache = new LRUCache[(String, DateTime, String), BinaryJobInfo](maxEntries)
   private val logger = LoggerFactory.getLogger(getClass)
   implicit val daoAskTimeout: Timeout = Timeout(3 seconds)
 
   /**
    * Retrieves the given SparkJob class from the cache if it's there, otherwise use the DAO to retrieve it.
-   * @param appName the appName under which the jar was uploaded
-   * @param uploadTime the upload time for the version of the jar wanted
+   * @param appName the appName under which the binary was uploaded
+   * @param uploadTime the upload time for the version of the binary wanted
    * @param classPath the fully qualified name of the class/object to load
    */
   def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
-    cache.get((appName, uploadTime, classPath), {
-      import akka.pattern.ask
-      import scala.concurrent.Await
-
-      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
-      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
-      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
-      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
-      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
-      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
-      JobJarInfo(constructor, classPath, jarFilePath)
-    })
+    cache.get((appName, uploadTime, classPath)) match {
```

Comment:
Ideally, we should rewrite the cache entirely to be future based and just pass futures around on this kind of stuff.